PWGHF/HFC/Tasks/taskCorrelationD0.cxx

```diff
@@ -0,0 +1,501 @@
+    }
+    if (cutYCandMax >= 0. && std::abs(YD0(candidate)) > cutYCandMax) {
+      return false;
+    }
+    return true;
+  }
+
+  //  TODO: Note: we do not need all these plots since they are in D0 and Lc task -> remove it after we are sure this works
+  template <typename TTracks>
+  void fillCandidateQA(TTracks candidates)
+  {
+    for (auto& candidate : candidates) {
+      if (!isAcceptedCandidate(candidate)) {
+        continue;
+      }
+
+      if (candidate.isSelD0() >= d_selectionFlagD0) {
+        registry.fill(HIST("hMass"), InvMassD0(candidate), candidate.pt());
+      }
+      if (candidate.isSelD0bar() >= d_selectionFlagD0bar) {
+        registry.fill(HIST("hMass"), InvMassD0bar(candidate), candidate.pt());
+      }
+
+      registry.fill(HIST("hPtCand"), candidate.pt());
+      registry.fill(HIST("hPtProng0"), candidate.ptProng0());
+      registry.fill(HIST("hPtProng1"), candidate.ptProng1());
+      registry.fill(HIST("hDecLength"), candidate.decayLength(), candidate.pt());
+      registry.fill(HIST("hDecLengthXY"), candidate.decayLengthXY(), candidate.pt());
+      registry.fill(HIST("hd0Prong0"), candidate.impactParameter0(), candidate.pt());
+      registry.fill(HIST("hd0Prong1"), candidate.impactParameter1(), candidate.pt());
+      registry.fill(HIST("hd0d0"), candidate.impactParameterProduct(), candidate.pt());
+      registry.fill(HIST("hCTS"), CosThetaStarD0(candidate), candidate.pt());
+      registry.fill(HIST("hCt"), CtD0(candidate), candidate.pt());
+      registry.fill(HIST("hCPA"), candidate.cpa(), candidate.pt());
+      registry.fill(HIST("hEtaCand"), candidate.eta(), candidate.pt());
+      registry.fill(HIST("hSelectionStatus"), candidate.isSelD0() + (candidate.isSelD0bar() * 2), candidate.pt());
+      registry.fill(HIST("hImpParErr"), candidate.errorImpactParameter0(), candidate.pt());
+      registry.fill(HIST("hImpParErr"), candidate.errorImpactParameter1(), candidate.pt());
+      registry.fill(HIST("hDecLenErr"), candidate.errorDecayLength(), candidate.pt());
+      registry.fill(HIST("hDecLenXYErr"), candidate.errorDecayLengthXY(), candidate.pt());
+    }
+  }
+
+  template <typename TTarget, typename TTracksTrig, typename TTracksAssoc>
+  void fillCorrelations(TTarget target, TTracksTrig tracks1, TTracksAssoc tracks2, float multiplicity, float posZ)
+  {
+    auto triggerWeight = 1;
+    auto associatedWeight = 1;
+
+    for (auto& track1 : tracks1) {
+
+      float eta1 = track1.eta();
+      float pt1 = track1.pt();
+      float phi1 = track1.phi();
+      o2::math_utils::bringTo02Pi(phi1);
+
+      //  TODO: add getter for NUE trigger efficiency here
+
+      //  calculating inv. mass to be filled into the container below
+      //  Note: this is needed only in case of HF-hadron correlations
+      bool fillingHFcontainer = false;
+      float invmass = 0;
+      if constexpr (std::is_same_v<hfCandidates, TTracksTrig>) {
+        fillingHFcontainer = true;
+        invmass = InvMassD0(track1);
```

Comment:
Here it seems that you are considering candidates passing both D0 and D0bar selections (at least, this is the seleciton you apply at L283-287), but filling them always with the invariant mass corresponging to the D0 hypothesis.
If you want to consider bith D0 and D0bar, you should check then separately, for each candidate, whether it passes the D0 selection (and in that case use InvMassD0(track1)) and/or the D0bar selection (and in that case use InvMassD0bar(track1)). Note that a given candidate can pass both selections (in case of reflections), and in that situation it has to be filled twice (both in the containers, and in the mass plots). Of course, in this situation, one of the two entries entry will for sure be of background (but you don't know which, and it can be also both).